core-bundle/src/Framework/ContaoFramework.php

```diff
@@ -158,26 +168,36 @@ public function getAdapter($class): Adapter
      */
     private function setConstants(): void
     {
-        if (!\defined('TL_MODE')) {
+        if (null !== $this->request && !\defined('TL_MODE')) {
             \define('TL_MODE', $this->getMode());
         }
 
-        \define('TL_START', microtime(true));
-        \define('TL_ROOT', $this->rootDir);
-        \define('TL_REFERER_ID', $this->getRefererId());
+        if (!\defined('TL_START')) {
+            \define('TL_START', microtime(true));
+        }
+
+        if (!\defined('TL_ROOT')) {
```

Comment:
This check should not have to be there, as the framework cannot be initialized twice. Did you add it for the unit tests only?

Reply:
`TL_ROOT` cannot be initialized if the request is not matched.